File: sdk/STOMP/Ruby/consumer.rb
# File: consumer.rb
# Description: This is the STOMP consumer handles incoming messages
#     published by producers from a particular queue.
#     Consumer prints the message body as it receives messages.

# Author: Wesley Zhang
# robomq.io (http://www.robomq.io)

require 'stomp'

# connection options
server = "hostname"
port = "61613"
vhost = "yourvhost"
login = "username"
passcode = "password"
destination = "/queue/test"

# stomp gem connect hash
hash = { :hosts => [
  {:login => login, :passcode => passcode, :host => server, :port => port},
  ],
  :connect_headers => {"host" => vhost, "accept-version" => "1.2", "heart-beat" => "60000,0", "content-type": "text/plain"}
}

loop do
  begin
    # connect
    connection = Stomp::Connection.new(hash)

    # subscribe
    connection.subscribe(destination, {"ack": "client-individual", "id": "0"})
    while msg = connection.receive
      puts msg.body
      # ack current message
      connection.ack(msg.headers['message-id'])
    end
  rescue => e
    puts "Exception handled, reconnecting...\nDetail:\n#{e.message}"
    sleep 5
  end
end

File: sdk/STOMP/Go/consumer.go
package main

import (
	"fmt"
	"github.com/go-stomp/stomp"
	"net"
	"time"
)

var server = "hostname"
var port = "61613"
var vhost = "yourvhost"
var login = "username"
var passcode = "password"
var destination = "/queue/test" // There're more options other than /queue/...

func main() {
	// Infinite loop to auto-reconnect on failure
Loop:
	for {
		fmt.Println("Starting in 5 seconds...")
		time.Sleep(5 * time.Second)

		// Connect to broker
		client, err := stomp.Dial("tcp", net.JoinHostPort(server, port),
			stomp.ConnOpt.Login(login, passcode),
			stomp.ConnOpt.Host(vhost),
			stomp.ConnOpt.AcceptVersion(stomp.V12),
			stomp.ConnOpt.HeartBeat(60*time.Second, 0*time.Second))
		if err != nil {
			fmt.Printf("Failed to connect, err: %v\n", err)
			continue Loop
		}
		// Subscribe to queue with client acknowledgement
		sub, err := client.Subscribe(destination, stomp.AckClient)
		if err != nil {
			fmt.Printf("Failed to subscribe, err: %v\n", err)
			continue Loop
		}
		for {
			msg := <-sub.C
			if msg.Err != nil {
				fmt.Printf("Can't handle message received, NACKing... Error: %v\n", msg.Err)
				// Unacknowledge the message
				err = client.Nack(msg)
				if err != nil {
					fmt.Printf("Failed to NACK, err: %v\n", err)
					break
				}
			}

			fmt.Println(string(msg.Body))
			// Acknowledge the message
			err = client.Ack(msg)
			if err != nil {
				fmt.Printf("Failed to ACK, err: %v\n", err)
				break
			}
		}
	}
}